### ISSUE TITLE
Type of enum variable has wrong infered value in comparison (false positive TS2367)

### ISSUE THREAD

**Issue by serkonda7** [microsoft/TypeScript#52776]
# Bug Report

### 🔎 Search Terms
- TS2367

### 🕗 Version & Regression Information

- version: `4.9.5`
- same for `next`

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/C4TwDgpgBAKg9gawgOygXigbwFAEgECWyAJgFyyIoDSRxeAbgIYA25AzsAE5EDm2AvtmwoArgFsKSZDRJY8I5AmRwA7sgA0eCHABmm3MkZiI+pCAD6jNqYgWCYsHE7BNg7AGNmVtlACSDp2AcAyMIdi5ePBYCK3DuZD5cdzhkDk4Rd2AnAApDYzjedSho2Kg03gBKYNxgAAsCNgA6POgMFrwa+qaSnwwevEE3T28oAAVGTjYITmCsqTZyeCkAbQBdPAJiAA9yZHEAI2m8MEY6goTjzgh6cznFymQ8O8kUPGQILeBbxHupISgAVBkqkuBkspxsnMUAsXsg1kUTmcyhEElUOnUGo0oal0FBsWx0V1GpstriAAyEzGI2q46kDIS4E6TCDmeyOZxsbIVcj+dnANbVZgQYBQNmBGG8wJrXFrDoqepCyFEuaNQiyNAYJbUWiq2ysgLONG4Y0YprvT5cjpJFIcKAtXGmxofSCZcwtS3G3BCkU9XHtT0EHRQJWYlVq4joTUPGTEXUWKxGz2O83AD2e30YR3OiCu90VK2CAMG4BNMAiNi1XIQFR+Yu5UJFHoVfPGwu4K7AEScVBijn0rRbF2p7OZCBkWExxOBkNNMO0SNQEfAMeJzqYlNp-iL5hTOSe4FsOBCxrMOA8bIAcmmnCcF5buELblwS7doS55x41Szg5zqa10h1Fp7w7LtUEdMArhuFUmGYfsDA+VNE3AyDvgQB1lUQSlZ0QdD1wQ1CsOafC5lw7D5mWR0SXWE0iRJABqOj6UEIA)

### 💻 Code

```ts
type Token = {
	kind: TokenKind
	val: string
}

enum TokenKind {
	unknown,
	eof,
	name,
	key_as,
	key_import,
}

class Import{
	name: string
	alias: string
	constructor(name: string, alias: string){
		this.name = name
		this.alias = alias
	}
}

class Parser{
	tokens: Token[]
	idx: number
	path: string
	prev_tok: Token
	tok: Token
	next_tok: Token

    constructor(tokens: Token[], path: string){
		this.tokens = tokens
		this.idx = 0
		this.path = path
	}

	parse_imports(): Import[]{
		let imports: Import[] = []
		while(this.tok.kind == TokenKind.key_import){
			this.next()
			const name = this.expect_name()
			let alias = name
                        // bug is here:
			if (this.tok.kind == TokenKind.key_as){
				this.next()
				alias = this.expect_name()
			}
			imports.push(new Import(name, alias))
		}
		return imports
	}

	expect(expected: TokenKind){
		if(this.tok.kind == expected){
			this.next()
		} else {
			console.log('error')
		}
	}

	expect_name(): string{
		this.expect(TokenKind.name)
		return this.prev_tok.val
	}

	next(){
		this.prev_tok = this.tok
		this.tok = this.next_tok
		this.next_tok = this.tokens[this.idx]
		this.idx++
	}
}

```

### 🙁 Actual behavior
The compiler thinks TOkenKind must be key_import. 
```
src/parser.ts:52:8 - error TS2367: This comparison appears to be unintentional because the types 'TokenKind.key_import' and 'TokenKind.key_as' have no overlap.

52    if (this.tok.kind == TokenKind.key_as){
          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
> Note: line numbers do not match

### 🙂 Expected behavior

The token and thus TokenKind changes in `next` and `expect_name`, so it can have other values too.
The compiler shouldn't infer a fixed Type here.

**Comment by MartinJohns** [comment]
Duplicate of #9998.

**Comment by jakebailey** [comment]
FWIW TypeScript's own scanner gets around this by making a token function (or for you, method) which returns the token, then always uses that, thus stopping it from being narrowed.
